### PR TITLE
Add optional manual commit to seldon kafka server

### LINKED ISSUE #4097: Data loss from seldon deployment with kafka serverType and autoscaling
## Describe the bug

`Seldon` deployments with `kafka` `seldon` is causing data loss during scale-down events. The scenario is a model which takes around 10 seconds to do inference, wrapped in a container with only 1 worker and 1 thread. So the throughput of one replica is 0.1 inf/s. `Seldon` container engine is accepting inputs regardless of model's state.

The problem is the inputs are read by seldon container engine, and commited by the `kafka` consumer, even though they are not processed yet. So if a kill signal is received on the deployment, the data points that are received by the executor, but not yet processed are lost. This can be resolved by adding an option to disable automatic committing and explicitly commit after the input has been processed.

## To reproduce

Build the image with the given `Dockerfile` below. Deploy `seldon-core-operator` with version 1.13.1 via helm. Send some requests via the `kafka_producer` script given in seldon examples. During scale down; i.e from n to n-1 or 1 to 0; if one replica gets terminated and the `seldon-container-engine` kafka consumer has committed the messages (since auto commit is being used by default), the messages that were ready for processing, but were not processed yet are going to be lost.

## Expected behaviour

Seldon container engine commits the messages whenever the processing, i.e inference, is done. This way the messages that were not processed are retained in the queue.

## Environment

* Cloud Provider: GKE
* Kubernetes Cluster Version: v1.21.10-gke.2000
* Deployed Seldon System Images: 
    - docker.io/seldonio/seldon-core-operator:1.13.1
    - docker.io/seldonio/seldon-core-operator@sha256:1aaa551e3017dee01183c4cbfaebe395f43364bbaf0114a710c7dade52a1eee1

## Model Details 
* Dockerfile:
```dockerfile
FROM python:3.8

RUN pip install --upgrade pip

RUN pip install seldon-core werkzeug==2.0.3

COPY . .

ENTRYPOINT seldon-core-microservice SeldonEntryClass --service-type MODEL --workers 1 
```
* `SeldonEntryClass.py`:
```python
import logging
from time import sleep

import numpy as np


class SeldonEntryClass:
    def __init__(self):
        logging.info("Initalizing the Model!")

    def predict(self, X, *args, **kwargs) -> np.ndarray:
        logging.warn(f"Ignoring X: {X}")
        sleep(10)
        return X

```
* `deployment.yaml`:
```yaml
apiVersion: machinelearning.seldon.io/v1
kind: SeldonDeployment
metadata:
  name: seldon-sleeper
  namespace: seldon
spec:
  transport: rest
  serverType: kafka
  predictors:
    - componentSpecs:
        - spec:
            containers:
              - image: <private_repo>/seldon_sleeper:latest
                name: seldon-sleeper-model
          kedaSpec:
            pollingInterval: 15
            minReplicaCount: 0 
            cooldownPeriod: 60
            maxReplicaCount: 3
            triggers:
              - type: kafka
                metadata:
                  bootstrapServers: seldon-kafka-bootstrap.kafka.svc.cluster.local:9093
                  consumerGroup: model.seldon-sleeper.seldon
                  lagThreshold: "5"
                  topic: seldon-sleeper-input
                  offsetResetPolicy: latest
      svcOrchSpec:
        env:
          - name: KAFKA_BROKER
            value: seldon-kafka-bootstrap.kafka.svc.cluster.local:9093
          - name: KAFKA_INPUT_TOPIC
            value: seldon-sleeper-input
          - name: KAFKA_OUTPUT_TOPIC
            value: seldon-sleeper-output
      graph:
        name: seldon-sleeper-model
        type: MODEL
      name: model
      replicas: 1
```

### PR DESCRIPTION
**Pull request by YmirKhang**
Signed-off-by: Emirhan Karagül <emirhan350z@gmail.com>

**What this PR does / why we need it**: This PR provides an optional flag to control auto commit behavior of kafka consumer. When no environment variable is set in the `svcOrchSpec`, the executor works as usual. If auto commit is disabled, the messages will be committed only after they have been processed and response has been sent to the output queue.

**Which issue(s) this PR fixes**:

Fixes #4097 

**Special notes for your reviewer**: As mentioned before, disabled auto commit might cause data duplication when `kafka_workers > 1`.

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **sachinvarghese**
You can assign the PR to them by writing `/assign @sachinvarghese` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: executor/api/kafka/server.go [hunk 1]
```diff
@@ -101,6 +104,11 @@ func NewKafkaServer(fullGraph bool, workers int, deploymentName, namespace, prot
 
 		}
 	}
+
+	if !autoCommit && workers > 1 {
+		log.Info("Disabling auto commit for kafka can have undesired side effects with multiple workers")
```
Comment (YmirKhang) [inline executor/api/kafka/server.go:109]: Warning here.

**Comment by YmirKhang** [comment]
@fg91

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[cliveseldon](https://github.com/SeldonIO/seldon-core/pull/4117#pullrequestreview-997639044 "Approved")*

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [cliveseldon]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by axsaucedo** [comment]
Nice one @YmirKhang thank you for the contribution 

/approve

**Pull request merged**